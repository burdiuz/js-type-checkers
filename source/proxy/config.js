export const config = {
  wrapFunctionReturnValues: true,
  wrapFunctionArguments: false,
  wrapSetPropertyValues: true,
};

export const setProxyConfig = (newConfig) => Object.assign(config, newConfig);

export const getProxyConfig = () => ({ ...config });
