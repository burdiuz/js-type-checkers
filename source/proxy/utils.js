/* eslint-disable import/prefer-default-export */

import { isWrappable } from '../utils';

export const getTypeCheckedChild = (wrapFn, info, name, value) => {
  if (!isWrappable(value)) {
    return value;
  }

  const childInfo = info.getChild(name);

  if (childInfo) {
    return wrapFn(value, childInfo);
  }

  return wrapFn(value, info.createChild(name));
};
