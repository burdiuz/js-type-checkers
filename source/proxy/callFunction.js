import { getTargetInfo } from '../target/info';

import {
  getProxyConfigValue,
  PROXY_WRAP_FUNCTION_ARGUMENTS,
  PROXY_WRAP_FUNCTION_RETURN_VALUES
} from './config';

import { getTypeCheckedChild } from './utils';

const getTargetArguments = (createFn, target, argumentsList) => {
  const info = getTargetInfo(target);

  if (getProxyConfigValue(PROXY_WRAP_FUNCTION_ARGUMENTS, info)) {
    const { length } = argumentsList;
    // FIXME cache arguments info objects as children
    for (let index = 0; index < length; index++) {
      argumentsList[index] = getTypeCheckedChild(
        createFn,
        info,
        String(index),
        argumentsList[index]
      );
    }
  }

  return argumentsList;
};

const callFunction = (createFn) => (target, thisArg, argumentsList) => {
  const info = getTargetInfo(target);
  const { names, config, checker } = info;

  if (checker.arguments) {
    checker.arguments(target, thisArg, argumentsList, config, names);
  }

  argumentsList = getTargetArguments(createFn, target, argumentsList);

  let result = target.apply(thisArg, argumentsList);

  if (checker.returnValue) {
    checker.returnValue(target, thisArg, result, config, names);
  }

  if (getProxyConfigValue(PROXY_WRAP_FUNCTION_RETURN_VALUES, info)) {
    result = getTypeCheckedChild(createFn, info, 'returnValue', result);
  }

  return result;
};

export default callFunction;
