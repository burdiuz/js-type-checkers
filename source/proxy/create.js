import { getDefaultTypeChecker } from '../checkers';
import { getErrorReporter } from '../reporters';
import { isEnabled } from '../enabled';
import {
  createTargetInfo,
  setTargetInfo,
  createChildrenCache,
} from '../target/info';
import { isValidTarget, isTypeChecked } from '../utils';
import getPropertyInit from './getProperty';
import setPropertyInit from './setProperty';
import callFunctionInit from './callFunction';

let getProperty;
let setProperty;
let callFunction;

const objectProxy = (target) => new Proxy(
  target,
  {
    get: getProperty,
    set: setProperty,
  },
);

const functionProxy = (target) => new Proxy(
  target,
  {
    apply: callFunction,
    construct: callFunction,
  },
);

export const wrapWithProxy = (target) => {
  if (target instanceof Function) {
    return functionProxy(target);
  }

  return objectProxy(target);
};

export const createInfoFromOptions = (target, {
  deep = true,
  names = [],
  config = null,
  children = null,
  checker = getDefaultTypeChecker(),
  info = null, // exclusive option, if set other options being ignored
} = {}) => info || createTargetInfo(
  checker,
  checker.init(target, getErrorReporter(), config),
  deep,
  names,
  createChildrenCache(children),
);

const create = (target, options) => {
  if (!isValidTarget(target) || !isEnabled() || isTypeChecked(target)) {
    return target;
  }

  setTargetInfo(
    target,
    createInfoFromOptions(target, options),
  );

  return wrapWithProxy(target);
};

getProperty = getPropertyInit(create);
setProperty = setPropertyInit(create);
callFunction = callFunctionInit(create);

export default create;
