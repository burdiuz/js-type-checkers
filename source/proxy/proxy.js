import { getDefaultTypeChecker } from '../checkers';
import { getErrorReporter } from '../reporters';
import { isEnabled } from '../enabled';
import {
  INFO_KEY,
  createTargetInfo,
  getTargetInfo,
  setTargetInfo,
  createChildrenCache,
  getChildInfo,
  storeChildInfoFrom,
  removeChildInfo,
} from '../target';
import { config as proxyConfig } from './config';
import { isValidTarget, isTypeChecked } from '../utils';
import { RETURN_VALUE } from '../checkers/utils';

const getProperty = (target, property) => {
  let value = target[property];

  if (property === INFO_KEY) {
    return value;
  }

  const info = getTargetInfo(target);
  const { deep, names, config, checker } = info;

  checker.getProperty
    && checker.getProperty(target, property, value, config, names);

  if (!isValidTarget(value) || isTypeChecked(value)) {
    return value;
  }

  if (deep || value instanceof Function) {
    const { children } = info;
    const childInfo = getChildInfo(children, name);

    if (childInfo) {
      value = create(value, { info: childInfo }, checker);
    } else {
      value = create(value, { deep, names: [...names, property] }, checker);
      storeChildInfoFrom(children, name, value);
    }
  }

  return value;
};

const setProperty = (target, property, value) => {
  const info = getTargetInfo(target);
  const { deep, names, config, checker } = info;

  if (property !== INFO_KEY) {
    checker.setProperty
      && checker.setProperty(target, property, value, config, names);

    if (proxyConfig.wrapSetPropertyValues) {
      const { children } = info;

      if (!isTypeChecked(value)) {
        const childInfo = getChildInfo(children, name);

        if (childInfo) {
          value = create(value, { info: childInfo }, checker);
        } else {
          value = create(value, { deep, names: [...names, property] }, checker);
        }
      }

      storeChildInfoFrom(children, name, value);
    }
  }

  target[property] = value;
};

const callFunction = (target, thisArg, argumentsList) => {
  const info = getTargetInfo(target);
  const { deep, names, config, checker } = info;

  checker.arguments
    && checker.arguments(target, thisArg, argumentsList, config, names);

  if (proxyConfig.wrapFunctionArguments) {
    const { length } = argumentsList;
    for (let index = 0; index < length; index++) {
      argumentsList[index] = create(argumentsList[index], { deep, names: [...names, index] }, checker);
    }
  }

  let result = target.apply(thisArg, argumentsList);

  checker.returnValue
    && checker.returnValue(target, thisArg, result, config, names);

  if (proxyConfig.wrapFunctionReturnValues) {
    const { children } = info;

    if (!isTypeChecked(result)) {
      const childInfo = getChildInfo(children, RETURN_VALUE);

      if (childInfo) {
        result = create(result, { info: childInfo }, checker);
      } else {
        result = create(result, { deep, names: [...names] }, checker)
      }
    }

    storeChildInfoFrom(children, RETURN_VALUE, result);
  }
  return result;
};

const objectProxy = (target) => new Proxy(
  target,
  {
    get: getProperty,
    set: setProperty,
  },
);

const functionProxy = (target) => new Proxy(
  target,
  {
    apply: callFunction,
    construct: callFunction,
  },
);

export const create = (
  target,
  {
    deep = true,
    names = [],
    config = null,
    children = null,
    info = null, // exclusive option, if set other options being ignored
  } = {},
  checker = getDefaultTypeChecker(),
) => {
  if (!isValidTarget(target) || !isEnabled() || isTypeChecked(target)) {
    return target;
  }

  setTargetInfo(
    target,
    info || createTargetInfo(
      checker,
      checker.init(target, getErrorReporter(), config),
      deep,
      names,
      createChildrenCache(children),
    ),
  );

  if (target instanceof Function) {
    return functionProxy(target);
  }

  return objectProxy(target);
};

export const createDeep = () => {
  // FIXME add new factory function createDeep, it will have deep == true by default and init type checkers for all internal objects 
  // and functions by reassigning original values with type checked proxies
  // when creating checkers for children objects should check cached info
};