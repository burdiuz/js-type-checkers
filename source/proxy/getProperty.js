import {
  INFO_KEY,
  getTargetInfo,
  getChildInfo,
  storeChildInfoFrom,
} from '../target/info';
import { TARGET_KEY } from '../target/proxy';
import { isValidTarget, isTypeChecked } from '../utils';

const getTargetProperty = (createFn, target, property, value) => {
  const info = getTargetInfo(target);
  const { deep, children, names, checker } = info;

  if (deep || value instanceof Function) {
    const childInfo = getChildInfo(children, property);

    if (childInfo) {
      value = createFn(value, { info: childInfo });
    } else {
      value = createFn(value, { deep, names: [...names, property], checker });
      storeChildInfoFrom(children, property, value);
    }
  }

  return value;
};

const getProperty = (createFn) => (target, property) => {
  const value = target[property];

  if (property === INFO_KEY) {
    return value;
    /*
    target[TARGET_KEY] is a virtual property accessing which indicates
    if object is wrapped with type checked proxy or not.
    */
  } else if (property === TARGET_KEY) {
    return target;
  }

  const info = getTargetInfo(target);
  const { names, config, checker } = info;

  if (checker.getProperty) {
    checker.getProperty(target, property, value, config, names);
  }

  if (!isValidTarget(value) || isTypeChecked(value)) {
    return value;
  }

  return getTargetProperty(createFn, target, property, value);
};

export default getProperty;
