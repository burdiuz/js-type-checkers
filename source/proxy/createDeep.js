import { isEnabled } from '../enabled';
import {
  setTargetInfo,
  storeChildInfo,
  getChildInfo,
} from '../target/info';
import { isTypeChecked } from '../utils';
import { wrapWithProxy, createInfoFromOptions } from './create';

const deepInitializer = (target, options) => {
  const info = createInfoFromOptions(target, options);
  const { deep, names, checker, config, children } = info;

  Object.keys(target)
    .forEach((name) => {
      const value = target[name];

      checker.getProperty(target, name, value, config, names);

      // skip functions/methods since we get info about them only when being executed
      if (typeof value === 'object') {
        let childInfo = getChildInfo(children, name);

        if (childInfo) {
          deepInitializer(value, { info: childInfo });
        } else {
          childInfo = deepInitializer(value, { deep, names: [...names, name], checker });
          storeChildInfo(children, name, childInfo);
        }
      }
    });

  setTargetInfo(target, info);

  return info;
};

const createDeep = (target, options) => {
  if (!target || typeof target !== 'object' || !isEnabled() || isTypeChecked(target)) {
    return target;
  }

  deepInitializer(target, options);

  return wrapWithProxy(target);
};

export default createDeep;
