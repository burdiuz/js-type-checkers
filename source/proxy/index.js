import { getProxyConfig, setProxyConfig } from './config';
import create from './create';
import createDeep from './createDeep';

export {
  getProxyConfig,
  setProxyConfig,
  create,
  createDeep,
};
