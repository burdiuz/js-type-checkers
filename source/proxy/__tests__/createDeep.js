import createDeep from '../createDeep';
import { setDefaultTypeChecker } from '../../checkers';
import { getTargetInfo } from '../../target/info';
import { getOriginalTarget } from '../../target/proxy';
import { setEnabled } from '../../enabled';
import { isTypeChecked } from '../../utils';
import create from '../create';

const mockChecker = () => ({
  init: (target) => ({ type: `type-checker-${typeof target}` }),
  getProperty: () => true,
  setProperty: () => true,
  arguments: () => true,
  returnValue: () => true,
  mergeConfigs: (config1, config2, names) => ({ config1, config2, names }),
});

describe('createDeep()', () => {
  let target;

  class MyClass {
    constructor() {
      this.stringValue = 'my string';
      this.booleanValue = 0;
      this.numbers = [0, 2, 3, 4];
    }
  }

  beforeEach(() => {
    setDefaultTypeChecker(mockChecker());
  });

  describe('When type checkers disabled', () => {
    beforeEach(() => {
      setEnabled(false);
      target = { list: [0, 1, 2, 3] };
    });

    afterEach(() => {
      setEnabled(true);
    });

    it('should return target as is', () => {
      expect(createDeep(target)).toBe(target);
    });

    it('should not be type checked', () => {
      expect(isTypeChecked(createDeep(target))).toBe(false);
      expect(isTypeChecked(createDeep(target).list)).toBe(false);
    });
  });

  describe('When target is not valid', () => {
    it('should return target as is', () => {
      const symbol = Symbol('1213223');
      expect(createDeep('my str')).toBe('my str');
      expect(createDeep(123)).toBe(123);
      expect(createDeep(true)).toBe(true);
      expect(createDeep(symbol)).toBe(symbol);
    });
  });

  describe('When target is type checked already', () => {
    beforeEach(() => {
      target = create({
        child: { val1: 1, val2: '2' },
      });
    });

    it('should return target as is', () => {
      expect(createDeep(target)).toBe(target);
    });

    it('should skip checking children', () => {
      expect(getOriginalTarget(createDeep(target)).child).toEqual({
        val1: 1,
        val2: '2',
      });
    });
  });

  describe('When target is valid and not type checked', () => {
    beforeEach(() => {
      const object = {
        stringValue: 'my string',
        booleanValue: true,
        list: [0, '2', true, [{ type: 'my objects' }]],
        method: (a, b, c) => (a + b) / c,
        child: new MyClass(),
      };

      target = createDeep(object);
    });

    it('should touch every descendant to generate full types map', () => {
      expect(target).toMatchSnapshot();
    });

    it('should contain type checked arrays', () => {
      expect(getTargetInfo(target.list)).toMatchSnapshot();
    });

    describe('When creating an object with already existing info', () => {
      let info;
      let newTarget;

      beforeEach(() => {
        info = getTargetInfo(target);

        const object = {
          stringValue: new MyClass(),
          booleanValue: true,
          list: (a, b, c) => (a + b) / c,
          method: [0, '2', true, [{ type: 'my objects' }]],
          child: 'my string',
        };

        newTarget = getOriginalTarget(createDeep(object, { info }));
      });

      it('should apply Function ino to "method"', () => {
        expect(getTargetInfo(newTarget.method).config.type).toBe('type-checker-function');
        expect(getTargetInfo(newTarget.method)).toMatchSnapshot();
      });

      it('should apply Array info to "list"', () => {
        expect(getTargetInfo(newTarget.list).config.type).toBe('type-checker-object');
        expect(getTargetInfo(newTarget.list)).toMatchSnapshot();
      });
    });
  });
});
