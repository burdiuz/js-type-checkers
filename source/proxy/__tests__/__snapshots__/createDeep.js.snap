// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createDeep() When target is valid and not type checked When creating an object with already existing info should apply Array info to "list" 1`] = `
Object {
  "checker": Object {
    "arguments": [Function],
    "getProperty": [Function],
    "init": [Function],
    "mergeConfigs": [Function],
    "returnValue": [Function],
    "setProperty": [Function],
  },
  "children": Object {
    "@3": Object {
      "checker": Object {
        "arguments": [Function],
        "getProperty": [Function],
        "init": [Function],
        "mergeConfigs": [Function],
        "returnValue": [Function],
        "setProperty": [Function],
      },
      "children": Object {
        "@0": Object {
          "checker": Object {
            "arguments": [Function],
            "getProperty": [Function],
            "init": [Function],
            "mergeConfigs": [Function],
            "returnValue": [Function],
            "setProperty": [Function],
          },
          "children": Object {},
          "config": Object {
            "type": "type-checker-object",
          },
          "deep": true,
          "names": Array [
            "list",
            "3",
            "0",
          ],
        },
      },
      "config": Object {
        "type": "type-checker-object",
      },
      "deep": true,
      "names": Array [
        "list",
        "3",
      ],
    },
  },
  "config": Object {
    "type": "type-checker-object",
  },
  "deep": true,
  "names": Array [
    "list",
  ],
}
`;

exports[`createDeep() When target is valid and not type checked When creating an object with already existing info should apply Function ino to "method" 1`] = `
Object {
  "checker": Object {
    "arguments": [Function],
    "getProperty": [Function],
    "init": [Function],
    "mergeConfigs": [Function],
    "returnValue": [Function],
    "setProperty": [Function],
  },
  "children": Object {
    "@3": Object {
      "checker": Object {
        "arguments": [Function],
        "getProperty": [Function],
        "init": [Function],
        "mergeConfigs": [Function],
        "returnValue": [Function],
        "setProperty": [Function],
      },
      "children": Object {
        "@0": Object {
          "checker": Object {
            "arguments": [Function],
            "getProperty": [Function],
            "init": [Function],
            "mergeConfigs": [Function],
            "returnValue": [Function],
            "setProperty": [Function],
          },
          "children": Object {},
          "config": Object {
            "type": "type-checker-object",
          },
          "deep": true,
          "names": Array [
            "method",
            "3",
            "0",
          ],
        },
      },
      "config": Object {
        "type": "type-checker-object",
      },
      "deep": true,
      "names": Array [
        "method",
        "3",
      ],
    },
  },
  "config": Object {
    "type": "type-checker-function",
  },
  "deep": true,
  "names": Array [
    "method",
  ],
}
`;

exports[`createDeep() When target is valid and not type checked should contain type checked arrays 1`] = `
Object {
  "checker": Object {
    "arguments": [Function],
    "getProperty": [Function],
    "init": [Function],
    "mergeConfigs": [Function],
    "returnValue": [Function],
    "setProperty": [Function],
  },
  "children": Object {
    "@3": Object {
      "checker": Object {
        "arguments": [Function],
        "getProperty": [Function],
        "init": [Function],
        "mergeConfigs": [Function],
        "returnValue": [Function],
        "setProperty": [Function],
      },
      "children": Object {
        "@0": Object {
          "checker": Object {
            "arguments": [Function],
            "getProperty": [Function],
            "init": [Function],
            "mergeConfigs": [Function],
            "returnValue": [Function],
            "setProperty": [Function],
          },
          "children": Object {},
          "config": Object {
            "type": "type-checker-object",
          },
          "deep": true,
          "names": Array [
            "list",
            "3",
            "0",
          ],
        },
      },
      "config": Object {
        "type": "type-checker-object",
      },
      "deep": true,
      "names": Array [
        "list",
        "3",
      ],
    },
  },
  "config": Object {
    "type": "type-checker-object",
  },
  "deep": true,
  "names": Array [
    "list",
  ],
}
`;

exports[`createDeep() When target is valid and not type checked should touch every descendant to generate full types map 1`] = `
Object {
  "booleanValue": true,
  "child": MyClass {
    "booleanValue": 0,
    "numbers": Array [
      0,
      2,
      3,
      4,
    ],
    "stringValue": "my string",
    Symbol(type-checkers::info): Object {
      "checker": Object {
        "arguments": [Function],
        "getProperty": [Function],
        "init": [Function],
        "mergeConfigs": [Function],
        "returnValue": [Function],
        "setProperty": [Function],
      },
      "children": Object {
        "@constructor": Object {
          "checker": Object {
            "arguments": [Function],
            "getProperty": [Function],
            "init": [Function],
            "mergeConfigs": [Function],
            "returnValue": [Function],
            "setProperty": [Function],
          },
          "children": Object {},
          "config": Object {
            "type": "type-checker-function",
          },
          "deep": true,
          "names": Array [
            "child",
            "constructor",
          ],
        },
        "@numbers": Object {
          "checker": Object {
            "arguments": [Function],
            "getProperty": [Function],
            "init": [Function],
            "mergeConfigs": [Function],
            "returnValue": [Function],
            "setProperty": [Function],
          },
          "children": Object {
            "@constructor": Object {
              "checker": Object {
                "arguments": [Function],
                "getProperty": [Function],
                "init": [Function],
                "mergeConfigs": [Function],
                "returnValue": [Function],
                "setProperty": [Function],
              },
              "children": Object {},
              "config": Object {
                "type": "type-checker-function",
              },
              "deep": true,
              "names": Array [
                "child",
                "numbers",
                "constructor",
              ],
            },
          },
          "config": Object {
            "type": "type-checker-object",
          },
          "deep": true,
          "names": Array [
            "child",
            "numbers",
          ],
        },
      },
      "config": Object {
        "type": "type-checker-object",
      },
      "deep": true,
      "names": Array [
        "child",
      ],
    },
  },
  "list": Array [
    0,
    "2",
    true,
    Array [
      Object {
        "type": "my objects",
        Symbol(type-checkers::info): Object {
          "checker": Object {
            "arguments": [Function],
            "getProperty": [Function],
            "init": [Function],
            "mergeConfigs": [Function],
            "returnValue": [Function],
            "setProperty": [Function],
          },
          "children": Object {
            "@constructor": Object {
              "checker": Object {
                "arguments": [Function],
                "getProperty": [Function],
                "init": [Function],
                "mergeConfigs": [Function],
                "returnValue": [Function],
                "setProperty": [Function],
              },
              "children": Object {},
              "config": Object {
                "type": "type-checker-function",
              },
              "deep": true,
              "names": Array [
                "list",
                "3",
                "0",
                "constructor",
              ],
            },
          },
          "config": Object {
            "type": "type-checker-object",
          },
          "deep": true,
          "names": Array [
            "list",
            "3",
            "0",
          ],
        },
      },
    ],
  ],
  "method": [Function],
  "stringValue": "my string",
  Symbol(type-checkers::info): Object {
    "checker": Object {
      "arguments": [Function],
      "getProperty": [Function],
      "init": [Function],
      "mergeConfigs": [Function],
      "returnValue": [Function],
      "setProperty": [Function],
    },
    "children": Object {
      "@child": Object {
        "checker": Object {
          "arguments": [Function],
          "getProperty": [Function],
          "init": [Function],
          "mergeConfigs": [Function],
          "returnValue": [Function],
          "setProperty": [Function],
        },
        "children": Object {
          "@constructor": Object {
            "checker": Object {
              "arguments": [Function],
              "getProperty": [Function],
              "init": [Function],
              "mergeConfigs": [Function],
              "returnValue": [Function],
              "setProperty": [Function],
            },
            "children": Object {},
            "config": Object {
              "type": "type-checker-function",
            },
            "deep": true,
            "names": Array [
              "child",
              "constructor",
            ],
          },
          "@numbers": Object {
            "checker": Object {
              "arguments": [Function],
              "getProperty": [Function],
              "init": [Function],
              "mergeConfigs": [Function],
              "returnValue": [Function],
              "setProperty": [Function],
            },
            "children": Object {
              "@constructor": Object {
                "checker": Object {
                  "arguments": [Function],
                  "getProperty": [Function],
                  "init": [Function],
                  "mergeConfigs": [Function],
                  "returnValue": [Function],
                  "setProperty": [Function],
                },
                "children": Object {},
                "config": Object {
                  "type": "type-checker-function",
                },
                "deep": true,
                "names": Array [
                  "child",
                  "numbers",
                  "constructor",
                ],
              },
            },
            "config": Object {
              "type": "type-checker-object",
            },
            "deep": true,
            "names": Array [
              "child",
              "numbers",
            ],
          },
        },
        "config": Object {
          "type": "type-checker-object",
        },
        "deep": true,
        "names": Array [
          "child",
        ],
      },
      "@constructor": Object {
        "checker": Object {
          "arguments": [Function],
          "getProperty": [Function],
          "init": [Function],
          "mergeConfigs": [Function],
          "returnValue": [Function],
          "setProperty": [Function],
        },
        "children": Object {},
        "config": Object {
          "type": "type-checker-function",
        },
        "deep": true,
        "names": Array [
          "constructor",
        ],
      },
      "@list": Object {
        "checker": Object {
          "arguments": [Function],
          "getProperty": [Function],
          "init": [Function],
          "mergeConfigs": [Function],
          "returnValue": [Function],
          "setProperty": [Function],
        },
        "children": Object {
          "@3": Object {
            "checker": Object {
              "arguments": [Function],
              "getProperty": [Function],
              "init": [Function],
              "mergeConfigs": [Function],
              "returnValue": [Function],
              "setProperty": [Function],
            },
            "children": Object {
              "@0": Object {
                "checker": Object {
                  "arguments": [Function],
                  "getProperty": [Function],
                  "init": [Function],
                  "mergeConfigs": [Function],
                  "returnValue": [Function],
                  "setProperty": [Function],
                },
                "children": Object {
                  "@constructor": Object {
                    "checker": Object {
                      "arguments": [Function],
                      "getProperty": [Function],
                      "init": [Function],
                      "mergeConfigs": [Function],
                      "returnValue": [Function],
                      "setProperty": [Function],
                    },
                    "children": Object {},
                    "config": Object {
                      "type": "type-checker-function",
                    },
                    "deep": true,
                    "names": Array [
                      "list",
                      "3",
                      "0",
                      "constructor",
                    ],
                  },
                },
                "config": Object {
                  "type": "type-checker-object",
                },
                "deep": true,
                "names": Array [
                  "list",
                  "3",
                  "0",
                ],
              },
              "@constructor": Object {
                "checker": Object {
                  "arguments": [Function],
                  "getProperty": [Function],
                  "init": [Function],
                  "mergeConfigs": [Function],
                  "returnValue": [Function],
                  "setProperty": [Function],
                },
                "children": Object {},
                "config": Object {
                  "type": "type-checker-function",
                },
                "deep": true,
                "names": Array [
                  "list",
                  "3",
                  "constructor",
                ],
              },
            },
            "config": Object {
              "type": "type-checker-object",
            },
            "deep": true,
            "names": Array [
              "list",
              "3",
            ],
          },
          "@constructor": Object {
            "checker": Object {
              "arguments": [Function],
              "getProperty": [Function],
              "init": [Function],
              "mergeConfigs": [Function],
              "returnValue": [Function],
              "setProperty": [Function],
            },
            "children": Object {},
            "config": Object {
              "type": "type-checker-function",
            },
            "deep": true,
            "names": Array [
              "list",
              "constructor",
            ],
          },
        },
        "config": Object {
          "type": "type-checker-object",
        },
        "deep": true,
        "names": Array [
          "list",
        ],
      },
      "@method": Object {
        "checker": Object {
          "arguments": [Function],
          "getProperty": [Function],
          "init": [Function],
          "mergeConfigs": [Function],
          "returnValue": [Function],
          "setProperty": [Function],
        },
        "children": Object {
          "@constructor": Object {
            "checker": Object {
              "arguments": [Function],
              "getProperty": [Function],
              "init": [Function],
              "mergeConfigs": [Function],
              "returnValue": [Function],
              "setProperty": [Function],
            },
            "children": Object {},
            "config": Object {
              "type": "type-checker-function",
            },
            "deep": true,
            "names": Array [
              "method",
              "constructor",
            ],
          },
        },
        "config": Object {
          "type": "type-checker-function",
        },
        "deep": true,
        "names": Array [
          "method",
        ],
      },
    },
    "config": Object {
      "type": "type-checker-object",
    },
    "deep": true,
    "names": Array [],
  },
}
`;
